element_zstack/volume.py

```diff
@@ -0,0 +1,261 @@
+        create_schema (bool): when True (default), create schema in the database if it
+                            does not yet exist.
+        create_tables (bool): when True (default), create schema tables in the database
+                             if they do not yet exist.
+        linking_module (str): A string containing the module name or module containing
+            the required dependencies to activate the schema.
+
+    Dependencies:
+    Tables:
+        Scan: A parent table to Volume
+    Functions:
+        get_volume_root_data_dir: Returns absolute path for root data director(y/ies) with
+            all volumetric data, as a list of string(s).
+        get_volume_tif_file: When given a scan key (dict), returns the full path to the
+            TIF file of the volumetric data associated with a given scan.
+    """
+
+    if isinstance(linking_module, str):
+        linking_module = importlib.import_module(linking_module)
+    assert inspect.ismodule(
+        linking_module
+    ), "The argument 'linking_module' must be a module's name or a module"
+
+    global _linking_module
+    _linking_module = linking_module
+
+    schema.activate(
+        schema_name,
+        create_schema=create_schema,
+        create_tables=create_tables,
+        add_objects=_linking_module.__dict__,
+    )
+
+
+# -------------------------- Functions required by the Element -------------------------
+
+
+def get_volume_root_data_dir() -> list:
+    """Fetches absolute data path to volume data directories.
+
+    The absolute path here is used as a reference for all downstream relative paths used in DataJoint.
+
+    Returns:
+        A list of the absolute path(s) to volume data directories.
+    """
+    root_directories = _linking_module.get_volume_root_data_dir()
+    if isinstance(root_directories, (str, Path)):
+        root_directories = [root_directories]
+
+    return root_directories
+
+
+def get_volume_tif_file(scan_key: dict) -> (str, Path):
+    """Retrieve the full path to the TIF file of the volumetric data associated with a given scan.
+    Args:
+        scan_key: Primary key of a Scan entry.
+    Returns:
+        Full path to the TIF file of the volumetric data (Path or str).
+    """
+    return _linking_module.get_volume_tif_file(scan_key)
+
+
+# --------------------------------------- Schema ---------------------------------------
+
+@schema
+class Volume(dj.Imported):
+    definition = """
+    -> Scan
+    ---
+    px_width: int # total number of voxels in x dimension
+    px_height: int # total number of voxels in y dimension
+    px_depth: int # total number of voxels in z dimension
+    depth_mean_brightness=null: longblob  # mean brightness of each slice across the depth (z) dimension of the stack
+    volume: longblob  # volumetric data - np.ndarray with shape (z, y, x)
+    """
+
+    def make(self, key):
+        vol_tif_fp = get_volume_tif_file(key)
+        volume_data = TiffFile(vol_tif_fp).asarray()
```

Comment:
Should we add a step to reshape this data before inserting it below? Or is the data already expected to have dimensions (z, y, x) when it is loaded here?

Reply:
It is already (z, y, x) here